lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -464,6 +475,60 @@ CFEPAnalyzer<FunctionOracle>::CFEPAnalyzer(llvm::Module &M,
   }
 }
 
+static void
+initializeCache(llvm::Module &M,
+                FunctionAnalysisResults &Properties,
+                const std::vector<llvm::GlobalVariable *> &ABIRegisters,
+                model::Binary &Binary) {
+  using namespace model;
+  using RegisterState = model::RegisterState::Values;
+
+  for (const auto &F : Binary.Functions) {
+    if (F.Type == model::FunctionType::Invalid)
+      continue;
+
+    MetaAddress EntryPC = F.Entry;
+    auto &Summary = Properties[EntryPC];
```

Comment:
Precisely because there's no `Summary` for that PC, we need to create a new entry in `Properties`. The `at` method just performs a look up, and will throw an out-of-range exception if the PC is not found (which is exactly this case).

Reply:
Also, we still don't want to make the map `FunctionsBucket` field public, that's the reason of the introduction of `operator[]`, which adds a new element in case it doesn't exists.

Reply:
I understand, but the alternative to `operator[]` is something like `insert`/`emplace`.
`operator[]` is a more sophisticated tool than we need and it pulls in `DefaultSummary` which scares me a bit.

Reply:
I see, but then it just a matters of changing name, since the `insert` method would still need to `try_emplace` a `DefaultSummary`, no? Also, now that we moved `initializeCache` as a member of the class, we may want to move `operator[]` / `insert` as a private member function.

Reply:
Why do you need `DefaultSummary`? Isn't a default-constructed summary enough?
`DefaultSummary` role is to give a "default" analysis results for the situations where the analysis has not been performed yet, which doesn't seem to be the case here.

`(*this)[...]` is not very nice :P

Reply:
Why do you need `DefaultSummary`? Isn't a default-constructed summary enough?
`DefaultSummary` role is to give a "default" analysis results for the situations where the analysis has not been performed yet, which doesn't seem to be the case here.

`(*this)[...]` is not very nice :P

Reply:
The thing with the whole `FunctionSummary` was to allow to select it at runtime, thus creating a default-constructed summary makes no sense from a s semantic point of view; indeed we had decided to delete it. As far as I see it, each entry PC should have associated a `DefaultSummary` or a summary, which is a more refined version of the default one, and never a default-constructed one. Even if the analysis won't be performed here, `initializeCache` will still bring `DefaultSummary` to a more accurate summary.

Reply:
More specifically, I reckon that, for the semantic the summary has (i.e., either a `DefaultSummary`, or an object returned from `milkInfo` using the parametrized constructor), one should never be allowed to default-instantiate a `FunctionSummary` object.

Reply:
Why do we care at all about `DefaultSummary` here? Is there something relevant we are inheriting from it?
If not, just use the constructor of `FunctionSummary`, even if it's not the default one.
What's the point of copying `DefaultSummary` just to clobber it immediately after everything was in it?

Reply:
Well, actually I just take advantage of the fact that `Summary.ClobberedRegisters` has been fulfilled with all the ABI regs (as is the case for `DefaultSummary`), and then we erase from the clobbered ones, those which are preserved.

Reply:
I think we are not on the same page on the role of `DefaultSummary`.
`DefaultSummary` role is to have a conservative, ABI-specific, summary for functions we know nothing about. It's not some sort of "starting point" for all the other summaries.

Right now, `DefaultSummary` it's being populated in a certain way, but it will become ABI-specific and out of your control.
You should not rely on its list of clobbered registers to be all the ABI registers.
If other parts of the code do that too, that's not nice either and it should be changed.

Reply:
Makes sense. What do you think now?